src/x509.rs

```diff
@@ -203,60 +204,35 @@ impl<A: SignatureAlgorithm + BERDecodable> BERDecodable for SubjectPublicKeyInfo
     }
 }
 
-/// Certificate `Version` as defined in [RFC 5280 Section 4.1].
-///
-/// ```text
-/// Version  ::=  INTEGER  {  v1(0), v2(1), v3(2)  }
-/// ```
-///
-/// [RFC 5280 Section 4.1]: https://datatracker.ietf.org/doc/html/rfc5280#section-4.1
-#[derive(Clone, Debug, Eq, PartialEq, Hash)]
-pub enum Version {
-    V1 = 1,
-    V2 = 2,
-    V3 = 3,
-}
-
-impl DerWrite for Version {
-    fn write(&self, writer: DERWriter) {
-        ((*self).clone() as u32).write(writer)
-    }
-}
-
-impl BERDecodable for Version {
-    fn decode_ber(reader: BERReader) -> ASN1Result<Self> {
-        let num = reader.read_u32()?;
-        match num {
-            1 => Ok(Version::V1),
-            2 => Ok(Version::V2),
-            3 => Ok(Version::V3),
-            _ => Err(ASN1Error::new(ASN1ErrorKind::Invalid)),
-        }
-    }
-}
-
-#[derive(Clone, Debug, Eq, PartialEq, Hash)]
-pub struct AttributeTypeAndValue {
-    pub oid: ObjectIdentifier,
-    pub value: TaggedDerValue,
-}
-
-impl DerWrite for AttributeTypeAndValue {
-    fn write(&self, writer: DERWriter) {
-        writer.write_sequence(|w| {
-            self.oid.write(w.next());
-            self.value.write(w.next());
-        });
+define_version! {
+    /// Certificate `Version` as defined in [RFC 5280 Section 4.1].
+    ///
+    /// ```text
+    /// Version  ::=  INTEGER  {  v1(0), v2(1), v3(2)  }
+    /// ```
+    ///
+    /// [RFC 5280 Section 4.1]: https://datatracker.ietf.org/doc/html/rfc5280#section-4.1
+    Version {
+        V1 = 1,
```

Comment:
I think these these should have values `0, 1, 2` instead of `1, 2, 3`.

Reply:
fixed